geth-utils/gethutil/trace.go

```diff
@@ -226,41 +230,115 @@ func Trace(config TraceConfig) ([]*ExecutionResult, error) {
 		stateDB.SetNonce(address, uint64(account.Nonce))
 		stateDB.SetCode(address, account.Code)
 		if account.Balance != nil {
-			stateDB.SetBalance(address, toBigInt(account.Balance))
+			stateDB.SetBalance(address, uint256.MustFromBig(toBigInt(account.Balance)), tracing.BalanceChangeUnspecified)
 		}
 		for key, value := range account.Storage {
 			stateDB.SetState(address, key, value)
 		}
 	}
 	stateDB.Finalise(true)
 
+	var (
+		err     error
+		usedGas uint64
+		raw     json.RawMessage
+		tx      types.Transaction
+	)
+
 	// Run the transactions with tracing enabled.
 	executionResults := make([]*ExecutionResult, len(config.Transactions))
 	for i, message := range messages {
 		tracer := logger.NewStructLogger(config.LoggerConfig)
-		evm := vm.NewEVM(blockCtx, core.NewEVMTxContext(&message), stateDB, &chainConfig, vm.Config{Debug: true, Tracer: tracer, NoBaseFee: true})
+		evm := vm.NewEVM(blockCtx, vm.TxContext{GasPrice: big.NewInt(0)}, stateDB, &chainConfig, vm.Config{Tracer: tracer.Hooks(), NoBaseFee: true})
+
+		tx = *types.NewTx(ToTxData(message, chainConfig.ChainID))
+		stateDB.SetTxContext(tx.Hash(), i)
 
-		result, err := core.ApplyMessage(evm, &message, new(core.GasPool).AddGas(message.GasLimit))
+		_, err = core.ApplyTransactionWithEVM(&message, &chainConfig, new(core.GasPool).AddGas(message.GasLimit), stateDB, blockCtx.BlockNumber, common.Hash{}, &tx, &usedGas, evm)
 		if err != nil {
-			executionResults[i] = &ExecutionResult{
-				Gas:         0,
-				Failed:      true,
-				Invalid:     true,
-				ReturnValue: fmt.Sprintf("%v", err),
-				StructLogs:  []StructLogRes{},
-			}
-		} else {
-			stateDB.Finalise(true)
+			return nil, fmt.Errorf("tracing failed: %w", err)
+		}
```

Comment:
I found that one of the testing failure is the invalid tx. This test contains two transactions: an invalid and then a valid one. I printed the execution result on main branch and it contains both GethExecTraces. The failing test on the PR branch contains only the invalid_tx's.

```
cargo test -p zkevm-circuits evm_circuit::execution::invalid_tx::test::invalid_tx_insufficient_gas -- --nocapture
```

The incorrect behavior happens at line 259. 
- Old behavior: When tx is invalid, keep the executionResult and continue processing the next tx.
- New behavior: When tx is invalid, early return the error message.

I modify the logic as below shows, the test can pass successfully.

```
		_, err = core.ApplyTransactionWithEVM(...)
		var result ExecutionResult
		if err != nil {
			result = ExecutionResult{
				Gas:         0,
				Failed:      true,
				Invalid:     true,
				ReturnValue: fmt.Sprintf("%v", err),
				StructLogs:  []StructLogRes{},
			}
		} else {
			raw, _ = tracer.GetResult()

			err = json.Unmarshal(raw, &result)
			if err != nil {
				return nil, fmt.Errorf("failed to unmarshal result: %w", err)
			}
		}

		executionResults[i] = &result
```

Reply:
Thank you very much!